subsys/usb/host/usbh_shell.c

```diff
@@ -560,54 +563,54 @@ static int cmd_usbh_disable(const struct shell *sh,
 }
 
 SHELL_STATIC_SUBCMD_SET_CREATE(desc_cmds,
-	SHELL_CMD_ARG(device, NULL, "<address>",
-		      cmd_desc_device, 2, 0),
-	SHELL_CMD_ARG(configuration, NULL, "<address> <index>",
-		      cmd_desc_config, 3, 0),
-	SHELL_CMD_ARG(string, NULL, "<address> <id> <index>",
-		      cmd_desc_string, 4, 0),
+	SHELL_CMD_ARG(device, NULL, NULL,
+		      cmd_desc_device, 1, 0),
+	SHELL_CMD_ARG(configuration, NULL, "<index>",
+		      cmd_desc_config, 2, 0),
+	SHELL_CMD_ARG(string, NULL, "<id> <index>",
+		      cmd_desc_string, 3, 0),
 	SHELL_SUBCMD_SET_END
 );
 
 SHELL_STATIC_SUBCMD_SET_CREATE(feature_set_cmds,
-	SHELL_CMD_ARG(rwup, NULL, "<address>",
-		      cmd_feature_set_rwup, 2, 0),
-	SHELL_CMD_ARG(ppwr, NULL, "<address> <port>",
-		      cmd_feature_set_ppwr, 3, 0),
-	SHELL_CMD_ARG(prst, NULL, "<address> <port>",
-		      cmd_feature_set_prst, 3, 0),
-	SHELL_CMD_ARG(halt, NULL, "<address> <endpoint>",
-		      cmd_feature_set_halt, 3, 0),
+	SHELL_CMD_ARG(rwup, NULL, NULL,
+		      cmd_feature_set_rwup, 1, 0),
+	SHELL_CMD_ARG(ppwr, NULL, "<port>",
+		      cmd_feature_set_ppwr, 2, 0),
+	SHELL_CMD_ARG(prst, NULL, "<port>",
+		      cmd_feature_set_prst, 2, 0),
+	SHELL_CMD_ARG(halt, NULL, "<endpoint>",
+		      cmd_feature_set_halt, 2, 0),
 	SHELL_SUBCMD_SET_END
 );
 
 SHELL_STATIC_SUBCMD_SET_CREATE(feature_clear_cmds,
-	SHELL_CMD_ARG(rwup, NULL, "<address>",
-		      cmd_feature_clear_rwup, 2, 0),
-	SHELL_CMD_ARG(halt, NULL, "<address> <endpoint>",
-		      cmd_feature_set_halt, 3, 0),
+	SHELL_CMD_ARG(rwup, NULL, NULL,
+		      cmd_feature_clear_rwup, 1, 0),
+	SHELL_CMD_ARG(halt, NULL, "<endpoint>",
+		      cmd_feature_set_halt, 2, 0),
 	SHELL_SUBCMD_SET_END
 );
 
 SHELL_STATIC_SUBCMD_SET_CREATE(device_cmds,
 	SHELL_CMD_ARG(address, NULL, "<address>",
 		      cmd_device_address, 2, 0),
-	SHELL_CMD_ARG(config, NULL, "<address> <config>",
-		      cmd_device_config, 3, 0),
-	SHELL_CMD_ARG(interface, NULL, "<address> <interface> <alternate>",
-		      cmd_device_interface, 4, 0),
+	SHELL_CMD_ARG(config, NULL, "<config>",
+		      cmd_device_config, 2, 0),
+	SHELL_CMD_ARG(interface, NULL, "<interface> <alternate>",
+		      cmd_device_interface, 3, 0),
 	SHELL_CMD_ARG(descriptor, &desc_cmds, "descriptor request",
 		      NULL, 1, 0),
 	SHELL_CMD_ARG(feature-set, &feature_set_cmds, "feature selector",
 		      NULL, 1, 0),
 	SHELL_CMD_ARG(feature-clear, &feature_clear_cmds, "feature selector",
 		      NULL, 1, 0),
-	SHELL_CMD_ARG(vendor_in, NULL, "<address> <length>",
-		      cmd_vendor_in, 3, 0),
-	SHELL_CMD_ARG(vendor_out, NULL, "<address> <length>",
-		      cmd_vendor_out, 3, 0),
-	SHELL_CMD_ARG(bulk, NULL, "<address> <endpoint> <length>",
-		      cmd_bulk, 4, 0),
+	SHELL_CMD_ARG(vendor_in, NULL, "<length>",
+		      cmd_vendor_in, 2, 0),
+	SHELL_CMD_ARG(vendor_out, NULL, "<length>",
+		      cmd_vendor_out, 2, 0),
+	SHELL_CMD_ARG(bulk, NULL, "<endpoint> <length>",
+		      cmd_bulk, 3, 0),
 	SHELL_SUBCMD_SET_END
 );
```

Comment:
Why are you removing `<address>` from the commands? Even with current just one device it is useful to be able to address packet to non-existent device to check whether the connected device really does not care about data addressed to others.

Reply:
I agree that we need this kind of commands or a way to test this request (must have).  But the address argument here is in the way to have device management. Thoughts: Inside struct usb_device there would be a path for the device, e.g. controller->hub->...->device. Now how to perform "row" request on the specific bus, one way is to have a dummy device with a path, the other is to submit "row" requests on specific controller bus. Anyway, this kind of commands should probably be for testing only and disabled by default. I am looking for a solution but unlikely it can be part or this PR.

Reply:
BUt how can you have device management if you cannot select the device you want to communicate with? In general USB is sending the data to all downstream ports, only the FS/LS devices connected via HS hub do not get other devices data (SPLIT token includes the port number where the FS/LS packets are supposed to go).

Because there can be only up to 5 hubs in between host and device, you can uniquely identify every device by knowing to which host controller it is connected to and the 5 bytes that form the path.

Reply:
But selecting a device, e.g. in the shell (in the future, it  would required some parts of device management be disabled) / submit a request to this device (the path to a device would be inside struct usb_device{}), and submit a specific request using a controller/bus are different use cases.